perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/AuthzResolverBlImpl.java

```diff
@@ -1922,4 +1964,172 @@ private static void prepareServiceRoles(PerunSession sess) {
+		}
+
+		return resolveAuthorization(sess, policyRoles, mapOfBeans);
+	}
+
+	/**
+	 * Decide whether a principal has sufficient rights according the the given roles and objects.
+	 *
+	 * @param sess perunSession which contains the principal.
+	 * @param policyRoles is a list of JsonNode from configuration file with a list of roles and objects they are connected with or another sub lists of them.
+	 * @param mapOfBeans is a map of objects against which will be authorization done.
+	 * @return true if the principal has particular rights, false otherwise.
+	 */
+	private static boolean resolveAuthorization(PerunSession sess, List<JsonNode> policyRoles, Map <String, List<PerunBean>> mapOfBeans) {
+		//Traverse through outer role list which works like logical OR
+		for (JsonNode roleArray: policyRoles) {
+
+			boolean authorized = true;
+			//Traverse through inner role list which works like logical AND
+			Iterator<String> roleArrayKeys = roleArray.fieldNames();
+			while (roleArrayKeys.hasNext()) {
+				String role = roleArrayKeys.next();
+
+				//fetch the role, object is connected with, from config
+				JsonNode roleObject = roleArray.get(role);
+
+				// If policy role is not connected to any object or if there is no corresponding type of object in the perunBeans map
+				if (roleObject.isNull() || !mapOfBeans.containsKey(roleObject.textValue())) {
+					//If principal does not have the role, this inner list's result is false
+					if (!sess.getPerunPrincipal().getRoles().hasRole(role)) {
+						authorized = false;
+						break;
+					}
+					// If policy role is connected to some object, like VOADMIN->Vo
+				} else {
+					//traverse all reliable objects from perun which are reliable for the authorized method
+					for (PerunBean object : mapOfBeans.get(roleObject.textValue())) {
+						//If the principal does not have rights on role-object, this inner list's result is false
+						if (!sess.getPerunPrincipal().getRoles().hasRole(role, object)) {
+							authorized = false;
+							break;
+						}
+					}
+					//Some inner role check failed so jump out of the loop
+					if (!authorized) break;
+				}
+			}
+			// If all checks for inner role list pass, return true. Otherwise proceed to another inner role list
+			if (authorized) return true;
+		}
+		//If all checks failed return false
+		return false;
+	}
+
+	/**
+	 * Fetch all possible PerunBeans for each of the objects from the list according to the id of the bean in the object.
+	 *
+	 *
+	 * @param objects for which will be reliable objects fetched.
+	 * @return all reliable objects together with the objects from the input.
+	 */
+	private static List<PerunBean> fetchAllReliableObjects(List<PerunBean> objects) {
+		List<PerunBean> reliableObjects = new ArrayList<>();
+
+		for (PerunBean object: objects) {
+			if (!reliableObjects.contains(object)) reliableObjects.add(object);
+			if (object.getBeanName().equals(Member.class.getSimpleName())) {
```

Comment:
What about rich objects? For example RichMember.

Reply:
Added.